### DISCUSSION TITLE
Prerelease support

### DISCUSSION THREAD

**Discussion by nbrugger-tgm** [gembaadvantage/uplift#53] [Ideas]
It is common for a Software lifecycle to Release prereleases or snapshot, dev or beta builds.

Therefore it would be Essential that a uplift supports such features to be used in this kind of Software.

Prereleases are a Part of the Semver spec.

`1.2.3-rc.12.5`
`1.2.3-beta`
`1.2.3-dev.264`

While as a User of semver you can use everything as pre build number (for example commit hash) but in the case of a standardized Software like uplift, simple increments should do the Trick. 

A few examples of how it would work (this build number Generation strategy is copied from the python comitizzen project) 
```
1.2.3 (breaking change, beta Release) 2.0.0-beta
1.2.3 (non breaking change, beta Release)
1.3.0-beta
1.2.3-beta (Patch, beta Release) 1.2.3-beta.2
1.2.3-beta (no changed, full release) 1.2.3
```
Be aware that I do not propose that this isn't the best way to handle it, as while it is convenient to have more Iteration of a Single beta Version that to have a new Version every change there are problems.

### Possible problems
 How would you Deal with this? 
```
v1.2.3
(minor changes beta Release)
v1.3.0-beta
(breaking change, beta Release)
??? v1.3.0-beta.2, v2.0.0-beta, v2.0.0-beta.2???
```
While I would say `2.0.0-beta` is the right answer It would t be quite easy to computer.

One would need to find out that the Jump from 1.2.3 to 1.3.0 is a "Feature" jump and therefore not sufficient to Cover a "breaking change jump.

> Another way simpler solution would be to just append whatever the User wants after the minus (as long as semver compilant) and dont give a ** about it for computing the next Version number

**Comment by ga-paul-t** [comment]
@nbrugger-tgm you have raised some valid points here. I will take a look at the `commitizzen` project. 

The initial idea behind `uplift` was to provide a mechanism for supporting conventional commits in a CI pipeline. Ideally with a focus towards continuous delivery/deployment, where the presence of a `feat!` (breaking change), `feat` or `fix` conventional commit would trigger a release i.e. by tagging the repository. For my personal workflow, when writing Go applications, the presence of a new tag would trigger goreleaser to build, package and distribute the binary.

In a way, it tries to encourage smaller, frequent commits by developers, or teams of developers.

But this may not always be the most ideal solution. And does alienate other ways of working. I'll dedicate some time to looking into the approach you have suggested above.

**Reply by nbrugger-tgm** [reply]
I think I would use uplift the same 😊 I would just use a different triggert: manual or branch merge.

And I also think for your intended workflow this Feature makes sense. Imagine you have a dev nranch with a working but not fully tested Version of uplift. You would not want to Tag them with Main Releases but want ppl to still be able to try New features?  Prerelease you go

**Reply by ga-paul-t** [reply]
I am leaning towards providing a mechanism through config that allows a user to specify a prerelease suffix like `beta` for example. The version bump will be calculated in exactly the same way, but the suffix will be appended. I'll raise a ticket for this feature shortly

**Reply by nbrugger-tgm** [reply]
i would rather do it using a command line argument as a file is way harder to controll in CI processes.
eg
its easy to do the following
`uplift bump --prerelease beta`
but kinda ridiculus to do
```shell
echo "var config = JSON.parse("""`cat uplift_config.json`""");" >> ~/temp.js
echo  "config.prerelease = \"beta \";" >> ~/temp.js 
echo "console.log(JSON.stringify(config));">>~/temp.js
echo $(node ~/temp.js) > uplift_config.json
```
Please don't make persons do this just to create a pre-release

Maybe i understand your intentions wrong but i would rather not do *If the config defines a pre-release just append it*

**Reply by nbrugger-tgm** [reply]
Also, this is not a rant but just ideas for improvement.

I think the main reason why a person would want to have pre-releases is a scenario where you want to create multiple breaking changes but not go from `2.3.4` to `7.0.0-rc` with `3-6.0.0-rc` "useless created. 

Because how i see it, with the current model you have in mind a mayor bump `2.3.4 to 3.0.0` can not have a pre-release because how i understood it, the workflow would go like this : <br/>
`2.3.4 -> 3.0.0-rc -> 4.0.0-rc -> 4.1.0`

**Reply by ga-paul-t** [reply]
Apologies think I misunderstood your initial use case. It would make much more sense to have this exposed as a CLI flag for running ad-hoc processes. I have been using uplift as a standalone process mainly without invoking individual parts of it through the different commands.

My long term target is to shrink the dependency on the config as much as possible. I would prefer the config to be an exception rather than the default behaviour (auto-detection of file types when bumping as an example). Which does contradict my previous comment :smile:

It will be easy to add a flag such as `--prerelease` as a global CLI flag that cascades through all of the other uplift commands

**Reply by ga-paul-t** [reply]
I am happy for any comments on possible improvements and ideas

**Reply by ga-paul-t** [reply]
Yes, you are right. The way uplift was designed initially was for a workflow of small frequent commits with each triggering a release that could then be cascaded into production through continuous deployment/delivery

**Reply by ga-paul-t** [reply]
Raised new issue to capture this, #65
